clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -0,0 +1,191 @@
+        this.principalBuilder = principalBuilder;
+    }
+
+    private SaslServer createSaslServer() throws IOException {
+        if (subject != null) {
+            // server is using a JAAS-authenticated subject: determine service principal name and hostname from kafka server's subject.
+            if (subject.getPrincipals().size() > 0) {
+                try {
+                    saslServerCallbackHandler = new SaslServerCallbackHandler(Configuration.getConfiguration());
+                    final Object[] principals = subject.getPrincipals().toArray();
+                    final Principal servicePrincipal = (Principal) principals[0];
+
+                    final String servicePrincipalNameAndHostname = servicePrincipal.getName();
+                    int indexOf = servicePrincipalNameAndHostname.indexOf("/");
+                    final String servicePrincipalName = servicePrincipalNameAndHostname.substring(0, indexOf);
+                    final String serviceHostnameAndKerbDomain = servicePrincipalNameAndHostname.substring(indexOf + 1, servicePrincipalNameAndHostname.length());
+                    indexOf = serviceHostnameAndKerbDomain.indexOf("@");
+                    final String serviceHostname = serviceHostnameAndKerbDomain.substring(0, indexOf);
+                    final String mech = "GSSAPI";
+
+                    LOG.debug("serviceHostname is '" + serviceHostname + "'");
+                    LOG.debug("servicePrincipalName is '" + servicePrincipalName + "'");
+                    LOG.debug("SASL mechanism(mech) is '" + mech + "'");
+                    boolean usingNativeJgss = Boolean.getBoolean("sun.security.jgss.native");
+                    if (usingNativeJgss) {
+                        try {
+                            GSSManager manager = GSSManager.getInstance();
+                            Oid krb5Mechanism = new Oid("1.2.840.113554.1.2.2");
+                            GSSName gssName = manager.createName(servicePrincipalName + "@" + serviceHostname, GSSName.NT_HOSTBASED_SERVICE);
+                            GSSCredential cred = manager.createCredential(gssName, GSSContext.INDEFINITE_LIFETIME, krb5Mechanism, GSSCredential.ACCEPT_ONLY);
+                            subject.getPrivateCredentials().add(cred);
+                        } catch (GSSException ex) {
+                            LOG.warn("Cannot add private credential to subject; clients authentication may fail", ex);
+                        }
+                    }
+
+                    try {
+                        return Subject.doAs(subject, new PrivilegedExceptionAction<SaslServer>() {
+                                public SaslServer run() {
+                                    try {
+                                        SaslServer saslServer;
+                                        saslServer = Sasl.createSaslServer(mech, servicePrincipalName, serviceHostname, null, saslServerCallbackHandler);
```

Comment:
Don't understand. What you mean by config properties in jaas config file. Jaas should only contain sections and it has specific syntax to them we shouldn't be treating it as generic config file.